onnx/shape_inference/shape_inference.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include "onnx/proto_utils.h"
+#include "onnx/defs/schema.h"
+
+namespace ONNX_NAMESPACE { namespace shape_inference {
+
+struct InferenceContextImpl : public InferenceContext {
+  virtual const AttributeProto* getAttribute(std::string name) {
+    auto attr = attributesByName_.find(name);
+    if (attr == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return &attr->second;
+    }
+  }
+  virtual const std::vector<TypeProto_Tensor>& getAllInputTypes() {
+    return allInputTypes_;
+  }
+  virtual Status setAllOutputTypes(std::vector<TypeProto_Tensor>&& allOutputTypes) {
+    allOutputTypes_ = allOutputTypes;
+    return Status::Success;
+  }
+  std::unordered_map<std::string, AttributeProto> attributesByName_;
```

Comment:
it can be const pointers to AttributeProto too

Reply:
++